### PR TITLE
UNR-1350: Only load & generate schema for map dependencies in GenerateSchemaAndSnapshots Commandlet

### PR DESCRIPTION
**Pull request by mattyoung-improbable**
#### Description
- Adds a new method to Schema Generator for finding all dependencies of a UWorld (including world composition deps).
- Add support for running schema generator to load a specific set of assets.
- Use the above changes to make GenerateSchemaAndSnapshot commandlet generate minimal schema for each map passed as an argument.

#### Release note
TODO

#### Tests
How did you test these changes prior to submitting this pull request?
What automated tests are included in this PR?

STRONGLY SUGGESTED: How can this be verified by QA?

#### Documentation
How is this documented (for example: release note, upgrade guide, feature page, in-code documentation)?

#### Primary reviewers
If your change will take a long time to review, you can name at most two primary reviewers who are ultimately responsible for reviewing this request. @ mention them.

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 1]
```diff
@@ -204,24 +215,38 @@ bool ValidateIdentifierNames(TArray<TSharedPtr<FUnrealType>>& TypeInfos)
...
 		{
-			UE_LOG(LogSpatialGDKSchemaGenerator, Warning, TEXT("Class name collision after removing non-alphanumeric characters. Name '%s' collides for types '%s' and '%s'. Will add a suffix to resolve the collision."),
-				*SchemaName, *Class->GetPathName(), *UsedSchemaNames[SchemaName]->GetPathName());
-
 			SchemaName = UnrealNameToSchemaName(ClassName) + FString::Printf(TEXT("%d"), ++Suffix);
 		}
 
-		ClassToSchemaName.Add(Class, SchemaName);
-		UsedSchemaNames.Add(SchemaName, Class);
+		ClassToSchemaName.Add(Class->GetPathName(), SchemaName);
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]:
```suggestion
		ClassToSchemaName.Add(ClassPath, SchemaName);
```

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 2]
```diff
@@ -204,24 +215,38 @@ bool ValidateIdentifierNames(TArray<TSharedPtr<FUnrealType>>& TypeInfos)
...
+		UsedSchemaNames.Add(SchemaName, Class->GetPathName());
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]:
```suggestion
		UsedSchemaNames.Add(SchemaName, ClassPath);
```

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 3]
```diff
@@ -204,24 +215,38 @@ bool ValidateIdentifierNames(TArray<TSharedPtr<FUnrealType>>& TypeInfos)
...
+	}
+
+	for (auto& Collision : NameCollisions)
+	{
+		if (Collision.Value.Num() > 1)
+		{
+			UE_LOG(LogSpatialGDKSchemaGenerator, Warning, TEXT("Class name collision after removing non-alphanumeric characters. Name '%s' collides for types [%s]"),
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Should we indicate that we resolved the collision by appending numbers and showing what maps to what now?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 4]
```diff
@@ -204,24 +215,38 @@ bool ValidateIdentifierNames(TArray<TSharedPtr<FUnrealType>>& TypeInfos)
...
 			SchemaName = UnrealNameToSchemaName(ClassName) + FString::Printf(TEXT("%d"), ++Suffix);
 		}
 
-		ClassToSchemaName.Add(Class, SchemaName);
-		UsedSchemaNames.Add(SchemaName, Class);
+		ClassToSchemaName.Add(Class->GetPathName(), SchemaName);
+		UsedSchemaNames.Add(SchemaName, Class->GetPathName());
+	}
+
+	for (auto& Collision : NameCollisions)
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]:
```suggestion
	for (const auto& Collision : NameCollisions)
```

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 5]
```diff
@@ -475,10 +500,24 @@ void TryLoadExistingSchemaDatabase()
...
 }
 
-bool SpatialGDKGenerateSchema()
+void ResetUsedNames()
 {
 	ClassToSchemaName.Empty();
 	UsedSchemaNames.Empty();
+	NameCollisions.Empty();
+
+	for(const TPair<FString, FSchemaData> Entry : ClassPathToSchema)
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]:
```suggestion
	for (const TPair<FString, FSchemaData>& Entry : ClassPathToSchema)
```

Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]:
```suggestion
	for(const TPair<FString, FSchemaData>& Entry : ClassPathToSchema)
```

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 6]
```diff
@@ -45,12 +45,23 @@ TMap<FString, uint32> LevelPathToComponentId;
 TSet<uint32> LevelComponentIds;
 
 // Prevent name collisions.
-TMap<UClass*, FString> ClassToSchemaName;
-TMap<FString, UClass*> UsedSchemaNames;
+TMap<FString, FString> ClassToSchemaName;
+TMap<FString, FString> UsedSchemaNames;
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: This is SchemaName to ClassName in FStrings right?. Could you add a comment to be obvious.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 7]
```diff
@@ -45,12 +45,23 @@ TMap<FString, uint32> LevelPathToComponentId;
...
+TMap<FString, TSet<FString>> NameCollisions;
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Same here, just a comment on what is stored.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp [hunk 8]
```diff
@@ -10,15 +10,41 @@
 #include "FileHelpers.h"
 
 #include "AssetRegistryModule.h"
+#include "GameFramework/GameModeBase.h"
+#include "GameFramework/WorldSettings.h"
 #include "GeneralProjectSettings.h"
+#include "Misc/ConfigCacheIni.h"
 #include "Misc/ScopedSlowTask.h"
 #include "UObject/StrongObjectPtr.h"
+#include "Engine/WorldComposition.h"
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp]: Please alphabetize

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 9]
```diff
@@ -45,12 +45,23 @@ TMap<FString, uint32> LevelPathToComponentId;
...
 
 namespace
 {
 
+void AddNameCollision(FString SchemaName, FString ClassPath)
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: This is for 'potential' collisions right? Maybe "AddSchemaNameUsage" or "AddUsedSchemaName"

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp [hunk 10]
```diff
@@ -113,15 +142,103 @@ bool FSpatialGDKEditor::GenerateSchema(bool bFullScan)
...
+			OutAssets.Add(NextDepAsset);
+			UE_LOG(LogSpatialGDKEditor, Verbose, TEXT("Adding %s(%s) to deps to load"), *NextDepAsset.GetFullName(), *NextDep.ToString());
+		}
+	}
+}
+
+bool FSpatialGDKEditor::LoadPotentialAssets(TArray<TStrongObjectPtr<UObject>>& OutAssetsLoaded, TSet<FAssetData> AssetsToLoad)
 {
 	FAssetRegistryModule& AssetRegistryModule = FModuleManager::LoadModuleChecked<FAssetRegistryModule>("AssetRegistry");
+	IAssetRegistry& AssetRegistry = AssetRegistryModule.Get();
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp]: Could this not be done inline with the `SearchAllAssets` call?

File: SpatialGDK/Source/SpatialGDKEditorCommandlet/Private/Commandlets/GenerateSchemaAndSnapshotsCommandlet.cpp [hunk 11]
```diff
@@ -166,33 +166,18 @@ bool UGenerateSchemaAndSnapshotsCommandlet::GenerateSchemaAndSnapshotForMap(FSpa
...
-	if (!GenerateSchemaForLoadedMap(InSpatialGDKEditor))
+	// Generate Schema
+	if (!GenerateSchemaForLoadedMap(InSpatialGDKEditor, AssetsToLoad))
 	{
 		return false;
 	}
 
+	AssetsAlreadyGenerated.Append(AssetsToLoad);
+
+
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDKEditorCommandlet/Private/Commandlets/GenerateSchemaAndSnapshotsCommandlet.cpp]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 12]
```diff
@@ -204,24 +215,38 @@ bool ValidateIdentifierNames(TArray<TSharedPtr<FUnrealType>>& TypeInfos)
...
+		if (!CheckSchemaNameValidity(SchemaName, ClassPath, TEXT("Class")))
 		{
 			bSuccess = false;
 		}
 
+		AddNameCollision(SchemaName, ClassPath);
+		
+		if (ClassToSchemaName.Contains(ClassPath))
+		{
+			continue;
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: What's this continue for?

Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: What's this continue for?

File: SpatialGDK/Source/SpatialGDKEditorCommandlet/Private/Commandlets/GenerateSchemaAndSnapshotsCommandlet.h [hunk 13]
```diff
@@ -25,10 +26,12 @@ class UGenerateSchemaAndSnapshotsCommandlet : public UCommandlet
...
 	TArray<FString> GeneratedMapPaths;
 
+	TSet<FAssetData> AssetsAlreadyGenerated;
+
 private:
 	bool GenerateSchemaAndSnapshotForPath(FSpatialGDKEditor& InSpatialGDKEditor, const FString& InPath);
 	bool GenerateSchemaAndSnapshotForMap(FSpatialGDKEditor& InSpatialGDKEditor, const FString& InMapName);
 
-	bool GenerateSchemaForLoadedMap(FSpatialGDKEditor& InSpatialGDKEditor);
+	bool GenerateSchemaForLoadedMap(FSpatialGDKEditor& InSpatialGDKEditor, TSet<FAssetData> AssetsToLoad);
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDKEditorCommandlet/Private/Commandlets/GenerateSchemaAndSnapshotsCommandlet.h]: Could AssetsToLoad be a reference?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 14]
```diff
@@ -204,24 +215,38 @@ bool ValidateIdentifierNames(TArray<TSharedPtr<FUnrealType>>& TypeInfos)
...
+		UsedSchemaNames.Add(SchemaName, Class->GetPathName());
+	}
+
+	for (auto& Collision : NameCollisions)
+	{
+		if (Collision.Value.Num() > 1)
+		{
+			UE_LOG(LogSpatialGDKSchemaGenerator, Warning, TEXT("Class name collision after removing non-alphanumeric characters. Name '%s' collides for types [%s]"),
+				*Collision.Key, *FString::Join(Collision.Value, TEXT(", ")));
+		}
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Maybe add a TODO here for handling class name collisions in an elegant way.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp (hunk 3 above)
Comment (Vatyx) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: I think we should.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 15]
```diff
@@ -475,10 +500,24 @@ void TryLoadExistingSchemaDatabase()
 	}
 }
 
-bool SpatialGDKGenerateSchema()
+void ResetUsedNames()
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Is this resetting or populating?

Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Is this resetting or populating?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/Utils/DataTypeUtilities.h [hunk 16]
```diff
@@ -6,7 +6,7 @@
 
 #include "SchemaGenerator/TypeStructure.h"
 
-extern TMap<UClass*, FString> ClassToSchemaName;
+extern TMap<FString, FString> ClassToSchemaName;
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/Utils/DataTypeUtilities.h]: Rename to ClassPathToSchemaName?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 17]
```diff
@@ -204,24 +215,38 @@ bool ValidateIdentifierNames(TArray<TSharedPtr<FUnrealType>>& TypeInfos)
...
+		const FString& ClassPath = Class->GetPathName();
 		FString SchemaName = UnrealNameToSchemaName(ClassName);
 
-		if (!CheckSchemaNameValidity(SchemaName, Class->GetPathName(), TEXT("Class")))
+		if (!CheckSchemaNameValidity(SchemaName, ClassPath, TEXT("Class")))
 		{
 			bSuccess = false;
 		}
 
+		AddNameCollision(SchemaName, ClassPath);
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Edge case, not really that important:
`(Unreal)A` -> `(schema)A`
`(Unreal)A!` -> collides with `(schema)A` -> `(schema)A1`
`(Unreal)A1` -> collides with `(schema)A1` -> `(schema)A11`
Here, `(Unreal)A!` won't be added in `AddNameCollision` for `(schema)A1`, only for `(schema)A`, so it won't log that `(Unreal)A1` had a collision.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/Utils/DataTypeUtilities.h (hunk 16 above)
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/Utils/DataTypeUtilities.h]: ClassNameToSchemaName / ClassPathToSchemaName ?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp [hunk 18]
```diff
@@ -44,8 +70,8 @@ bool FSpatialGDKEditor::GenerateSchema(bool bFullScan)
 
 	bSchemaGeneratorRunning = true;
 
-	// 80/10/10 load assets / gen schema / garbage collection.
-	FScopedSlowTask Progress(100.f, LOCTEXT("GeneratingSchema", "Generating Schema..."));
+	// Total Progress = Loading Assets (Assets) + Generating Schema (10) + Garbage Collecting (0.5 * Assets)
+	FScopedSlowTask Progress(AssetsToLoad.Num() * 1.5f + 10.f, LOCTEXT("GeneratingSchema", "Generating Schema..."));
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp]: hmmm yes scientific magic numbers

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/Utils/DataTypeUtilities.h (hunk 16 above)
Comment (Vatyx) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/Utils/DataTypeUtilities.h]: Should be ClassPath if what's being stored is the ClassPath.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp (hunk 8 above)
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp]: Sort includes.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp [hunk 19]
```diff
@@ -10,15 +10,41 @@
...
 #include "AssetRegistryModule.h"
+#include "GameFramework/GameModeBase.h"
+#include "GameFramework/WorldSettings.h"
 #include "GeneralProjectSettings.h"
+#include "Misc/ConfigCacheIni.h"
 #include "Misc/ScopedSlowTask.h"
 #include "UObject/StrongObjectPtr.h"
+#include "Engine/WorldComposition.h"
+
+
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp]: Extra whitespace?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp [hunk 20]
```diff
@@ -10,15 +10,41 @@
...
 
 DEFINE_LOG_CATEGORY(LogSpatialGDKEditor);
 
 #define LOCTEXT_NAMESPACE "FSpatialGDKEditor"
 
 bool FSpatialGDKEditor::GenerateSchema(bool bFullScan)
+{
+	FAssetRegistryModule& AssetRegistryModule = FModuleManager::LoadModuleChecked<FAssetRegistryModule>("AssetRegistry");
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp]: Is LoadModuleChecked safe to run, if it fails does this crash the editor?

Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp]: Also second place we've called this, perhaps make a function or part of an init?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp [hunk 21]
```diff
@@ -113,15 +142,103 @@ bool FSpatialGDKEditor::GenerateSchema(bool bFullScan)
 	return bResult;
 }
 
-bool FSpatialGDKEditor::LoadPotentialAssets(TArray<TStrongObjectPtr<UObject>>& OutAssets)
+void FSpatialGDKEditor::GetWorldDependencies(UWorld* World, TSet<FAssetData>& OutAssets, TSet<FAssetData>* AssetsToIgnore)
+{
+	TQueue<FName> DepsToSearch;
+	DepsToSearch.Enqueue(*World->GetOutermost()->GetPathName());
+
+	if (World->GetWorldSettings()->DefaultGameMode != nullptr)
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp]: Is this a GameMode override / is that override handled.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp [hunk 22]
```diff
@@ -113,15 +142,103 @@ bool FSpatialGDKEditor::GenerateSchema(bool bFullScan)
...
+
+			if (!GameModePath.IsEmpty())
+			{
+				UE_LOG(LogSpatialGDKEditor, Verbose, TEXT("Adding %s %s"), *GameMode, *FSoftObjectPath(GameModePath).GetLongPackageName());
+				DepsToSearch.Enqueue(*FSoftObjectPath(GameModePath).GetLongPackageName());
+			}
+		}
+	}
+
+	UE_LOG(LogSpatialGDKEditor, Verbose, TEXT("Loading all deps for %s"), *World->GetOutermost()->GetPathName());
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp]: "deps" = "dependencies"

Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp]: And rest of logs please.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp [hunk 23]
```diff
@@ -113,15 +142,103 @@ bool FSpatialGDKEditor::GenerateSchema(bool bFullScan)
...
+	if (World->WorldComposition != nullptr)
+	{
+		for (auto& Tile : World->WorldComposition->GetTilesList())
+		{
+			UE_LOG(LogSpatialGDKEditor, Verbose, TEXT("Adding World Comp Tile %s to Deps"), *Tile.PackageName.ToString());
+			DepsToSearch.Enqueue(Tile.PackageName);
+		}
+	}
+
+	FAssetRegistryModule& AssetRegistryModule = FModuleManager::LoadModuleChecked<FAssetRegistryModule>("AssetRegistry");
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp]: Yeah lets make this a function or a class object

Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp]: Yeah lets make this a function or a class object

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp [hunk 24]
```diff
@@ -113,15 +142,103 @@ bool FSpatialGDKEditor::GenerateSchema(bool bFullScan)
...
+	IAssetRegistry& AssetRegistry = AssetRegistryModule.Get();
+
+	TSet<FName> SearchedPackages;
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp]: Add a quick comment about what SearchedPackages is being used for, also we add `NextDep` to it meaning the name of dependency and package is a little confusing

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp [hunk 25]
```diff
@@ -136,13 +253,13 @@ bool FSpatialGDKEditor::LoadPotentialAssets(TArray<TStrongObjectPtr<UObject>>& O
...
 			return false;
 		}
-		Progress.EnterProgressFrame(1, FText::FromString(FString::Printf(TEXT("Loading %s"), *Data.AssetName.ToString())));
 		if (auto GeneratedClassPathPtr = Data.TagsAndValues.Find("GeneratedClass"))
 		{
 			const FString ClassObjectPath = FPackageName::ExportTextPathToObjectPath(*GeneratedClassPathPtr);
 			const FString ClassName = FPackageName::ObjectPathToObjectName(ClassObjectPath);
 			FSoftObjectPath SoftPath = FSoftObjectPath(ClassObjectPath);
-			OutAssets.Add(TStrongObjectPtr<UObject>(SoftPath.TryLoad()));
+			Progress.EnterProgressFrame(1.f, FText::FromString(FString::Printf(TEXT("Loading %s"), *Data.AssetName.ToString())));
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp]: This seems like the progress bar will not "finish" if there's any assets that aren't generated classes.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp [hunk 26]
```diff
@@ -113,15 +142,103 @@ bool FSpatialGDKEditor::GenerateSchema(bool bFullScan)
...
+			DepsToSearch.Enqueue(Tile.PackageName);
+		}
+	}
+
+	FAssetRegistryModule& AssetRegistryModule = FModuleManager::LoadModuleChecked<FAssetRegistryModule>("AssetRegistry");
+	IAssetRegistry& AssetRegistry = AssetRegistryModule.Get();
+
+	TSet<FName> SearchedPackages;
+
+	while(!DepsToSearch.IsEmpty())
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp]:
```suggestion
	while (!DepsToSearch.IsEmpty())
```

File: SpatialGDK/Source/SpatialGDKEditorCommandlet/Private/Commandlets/GenerateSchemaAndSnapshotsCommandlet.cpp [hunk 27]
```diff
@@ -166,33 +166,18 @@ bool UGenerateSchemaAndSnapshotsCommandlet::GenerateSchemaAndSnapshotForMap(FSpa
...
-
-	// Generate Schema Iteration
-	if (!GenerateSchemaForLoadedMap(InSpatialGDKEditor))
+	// Generate Schema
+	if (!GenerateSchemaForLoadedMap(InSpatialGDKEditor, AssetsToLoad))
 	{
 		return false;
 	}
 
+	AssetsAlreadyGenerated.Append(AssetsToLoad);
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditorCommandlet/Private/Commandlets/GenerateSchemaAndSnapshotsCommandlet.cpp]: Could an asset fail to load, just wondering.

**Comment by m-samiec** [comment]
This PR is stale, looks like everything went directly into release, close it up @mattyoung-improbable ?

**Comment by girayimprobable** [comment]
Did we end up implementing a stable way to get world dependencies? My impression was that we haven't yet.

**Comment by mattyoung-improbable** [comment]
> Did we end up implementing a stable way to get world dependencies? My impression was that we haven't yet.

The method in this PR turned out to be slower than using all assets. I think we should explore generation methods that don't involve loading assets at this point, either piggy backing on asset cooking, or building the data needed at asset save time.

**Pull request closed without merging**